crates/matrix-sdk/src/room_directory_search.rs

```diff
@@ -0,0 +1,370 @@
+            joined_members: 20,
+        }
+    }
+
+    async fn new_server_and_client() -> (MockServer, Client) {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri())).await;
+        (server, client)
+    }
+
+    #[async_test]
+    async fn search_success() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap();
+        let (results, _) = room_directory_search.results();
+        assert_eq!(results.len(), 1);
+        assert_eq!(results[0], get_first_page_description());
+        assert!(!room_directory_search.is_at_last_page);
+        assert_eq!(room_directory_search.loaded_pages(), 1);
+    }
+
+    #[async_test]
+    async fn search_success_paginated() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap();
+
+        Mock::given(RoomDirectorySearchMatcher {
+            next_token: Some("p190q".into()),
+            filter_term: None,
+            limit: 1,
+        })
+        .respond_with(
+            ResponseTemplate::new(200).set_body_json(&*test_json::PUBLIC_ROOMS_FINAL_PAGE),
+        )
+        .mount(&server)
+        .await;
+
+        room_directory_search.next_page().await.unwrap();
+
+        let (results, _) = room_directory_search.results();
+        assert_eq!(
+            results,
+            vec![get_first_page_description(), get_second_page_description()].into()
+        );
+        assert!(room_directory_search.is_at_last_page);
+        assert_eq!(room_directory_search.loaded_pages(), 2);
+    }
+
+    #[async_test]
+    async fn search_fails() {
+        let (server, client) = new_server_and_client().await;
+
+        let mut room_directory_search = RoomDirectorySearch::new(client);
+        Mock::given(RoomDirectorySearchMatcher { next_token: None, filter_term: None, limit: 1 })
+            .respond_with(ResponseTemplate::new(404))
+            .mount(&server)
+            .await;
+
+        room_directory_search.search(None, 1).await.unwrap_err();
+        let (results, _) = room_directory_search.results();
+        assert_eq!(results.len(), 0);
+        assert!(!room_directory_search.is_at_last_page);
+        assert_eq!(room_directory_search.loaded_pages(), 0);
```

Comment:
I think it's better to assert that `search()` returns an error instead of panicking with `unwrap_err()`.

```suggestion
        let search = room_directory_search.search(None, 1).await;
        assert!(search.is_err());

        let (results, stream) = room_directory_search.results();
        assert_eq!(results.len(), 0);
        assert!(!room_directory_search.is_at_last_page());
        assert_eq!(room_directory_search.loaded_pages(), 0);
        assert_pending!(stream);
```